### PR TITLE
chore(focus-trap): fix & improve the focus trap

### PR DESCRIPTION
**Pull request by gcornut**
- fix(focus-trap): fix reset focus trap
- fix(focus-trap): fix nested focus traps
- fix(focus-trap): fix close nested components on escape key pressed
- chore(popover): focus first focusable element inside anchor on close
- fix(date-picker-field): fixed focus element on open

# Test

Scenario:
1. Go to the story ["Dialog Focus Trap"](https://5fbfb1d508c0520021560f10-enyqcdxplj.chromatic.com/?path=/story/lumx-components-dialog-dialog--dialog-focus-trap)
1. Navigate with `Tab` to the "Open date picker" button
1. Open by pressing `Enter`
1. Navigate with `Tab` to the date picker field
1. Open by pressing `Enter`
1. Navigate dates with `Tab` (this navigation could be improved upon)
1. Close the date picker with `Escape`
1. Close the date picker dialog with `Escape`
1. Close the main dialog with `Escape`

Additional checks 
- Navigating with `Tab` or `shift+Tab` should be confined to the dialog/date-picker
- Closing a dialog/date-picker should focus the activating element (button or date-picker field)
- Opening a dialog/date-picker should focus an element by default

### CODE REVIEW

**Comment by mathieubrisson** [comment]
Hello @matmkian 

Could you have a look to this PR please ?
Thank you

File: packages/lumx-react/src/hooks/useCallbackOnEscape.ts [hunk 1]
```diff
@@ -1,25 +1,54 @@
...
+        // TEARDOWN:
+        return () => {
+            // Disable current listener.
+            listener.disable();
+            // Remove current listener.
+            pull(LISTENERS, listener);
+            // Enable previous listener.
+            last(LISTENERS)?.enable();
+        };
+    }, [callback, closeOnEscape]);
```
Comment (Mcourt) [inline packages/lumx-react/src/hooks/useCallbackOnEscape.ts:53]: I might be wrong, I don't have the full context and constraints in mind, but this code looks a lot like what you added in `useFocusTrap`.

Is there anyway we could reuse it somehow? Or maybe we voluntarily don't want to?

Comment (gcornut) [reply packages/lumx-react/src/hooks/useCallbackOnEscape.ts:53]: Good catch, I intended to mutualized but didn't find own to name this pattern I forgot to do it

**Comment by gcornut** [comment]
Rebased & factorized code

**Comment by Mcourt** [comment]
> Rebased & factorized code

Beautiful :')

**Pull request merged**